File: mengshi/topic_sub/src/sub_odom.cpp
//ros2 pkg create topic_sub --build-type ament_cmake --dependencies rclcpp nav_msgs --node-name sub_odom
#inclde "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/odometry.hpp"
using namespace std::placeholders;

class SubOdom : public rclcpp::Node
{
public:
    SubOdom() : Node("sub_odom")
    {
        last_x=0.0;
        last_y=0.0;
        this->declare_parameter<double>("distance", 0.5);
        sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "/odom", 10, std::bind(&SubOdom::on_timer, this, std::placeholders::_1));//_1
    }

private:
/*
    void on_timer(const nav_msgs::msg::Odometry::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "subscribe odom: [%f, %f, %f]",
                    msg->pose.pose.position.x,
                    msg->pose.pose.position.y,
                    msg->pose.pose.position.z);
    }
*/
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_;
    double last_x, last_y;
    void on_timer(const nav_msgs::msg::Odometry &odom)
    {
        double curr_x = odom.pose.pose.position.x;
        double curr_y = odom.pose.pose.position.y;
        //double curr_z = odom.pose.pose.position.z;
        double distance =sqrt(pow((curr_x-last_x),2)+pow((curr_y-last_y),2));
        if (distance >= this->get_parameter("distance").as_double())
        {
            RCLCPP_INFO(this->get_logger(), "subscribe odom: [%.2f, %.2f]",
                    curr_x, curr_y);
            last_x = curr_x;
            last_y = curr_y;
        }       
    }


}

int main(int argc, char **argv)//char const *argv[] is ok
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SubOdom>());
    rclcpp::shutdown();
    return 0;
}

File: mengshi/topic_sub_py/topic_sub_py/sub_odom.py
#ros2 pkg create topic_sub_py --build-type ament_python --dependencies rclpy nav_msgs --node-name sub_odom
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
import math
class SubOdom(Node):
    def __init__(self):
        super().__init__('sub_odom')
        self.last_x=0.0
        self.last_y=0.0
        self.subscription = self.create_subscription(
            Odometry,
            '/odom',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning
    def listener_callback(self, odom):
        x=odom.pose.pose.position.x
        y=odom.pose.pose.position.y
        dis_x=x-self.last_x
        dis_y=y-self.last_y
        distance=math.sqrt(dis_x*dis_x+dis_y*dis_y)
        if distance >= self.get_parameter('distance').get_parameter_value().double_value:
            self.get_logger().info('pos:(%.2f,%.2f)'%(x,y))
            self.last_x=x
            self.last_y=y
def main(args=None):
    rclpy.init(args=args)
    sub_odom = SubOdom()
    rclpy.spin(sub_odom)
    sub_odom.destroy_node()
    rclpy.shutdown()